torch/distributions/one_hot_categorical.py

```diff
@@ -74,11 +79,20 @@ def param_shape(self):
         return self._categorical.param_shape
 
     def sample(self, sample_shape=torch.Size()):
+        # work around the issue of `Tensor.expand` not respecting `torch.no_grad`
+        return self.rsample(sample_shape).detach()
```

Comment:
I think we should avoid the extra overhead of `.add(probs).sub(probs).detach()`. How about this refactoring:
```py
def sample(self, sample_shape=torch.Size()):
    # ...unchanged by this PR...

def rsample(self, sample_shape=torch.Size()):
    samples = self.sample(sample_shape)
    probs = self._categorical.probs  # note this is cached via @lazy_property
    return samples + (probs - probs.detach())
```

Reply:
This looks great and I'll refactor the code over the weekend.